### ISSUE TITLE
[0.88-SNAPSHOT] RealmTransformer should support both 1.5 and 2.0.0-beta

### ISSUE THREAD

**Issue by roughboy** [realm/realm-java#2348]
Using latest snapshot i get following error during gradle sync:
`Error:Unable to load class 'io.realm.transformer.RealmTransformer'.`

	buildscript {
		repositories {
			jcenter()
			maven {
				url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
			}
			maven {
				url "https://jitpack.io"
			}
		}
		dependencies {
			classpath 'io.realm:realm-gradle-plugin:0.88.0-SNAPSHOT'
		}
	}

	apply plugin: 'realm-android'

**Comment by cmelchior** [comment]
Hi @roughboy, Thank you for the report. We are aware of it and are working on fixing it asap.

**Comment by cmelchior** [comment]
Hi @roughboy 
The distribution should work now. Can you try again? 
Remember to clean you dependency cache using `./gradlew clean assemble --refresh-dependencies`

**Comment by roughboy** [comment]
Hi,
now RealmTransformer is gone but another one appeared, totally different. Below stacktrace from gradle build:

	FAILURE: Build failed with an exception.

	* What went wrong:
	com.android.build.api.transform.Transform.getSecondaryFiles()Ljava/util/Collection;

	* Try:
	Run with --info or --debug option to get more log output.

	* Exception is:
	java.lang.NoSuchMethodError: com.android.build.api.transform.Transform.getSecondaryFiles()Ljava/util/Collection;
		at com.android.build.gradle.internal.pipeline.TransformTask.getOtherFileInputs(TransformTask.java:78)
		at com.android.build.gradle.internal.pipeline.TransformTask_Decorated.getOtherFileInputs(Unknown Source)
		at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
		at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$PropertyInfo$4.create(AnnotationProcessingTaskFactory.java:458)
		at org.gradle.util.SingleMessageLogger.whileDisabled(SingleMessageLogger.java:166)
		at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$PropertyInfo.getValue(AnnotationProcessingTaskFactory.java:456)
		at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$Validator$1.call(AnnotationProcessingTaskFactory.java:263)
		at org.gradle.api.internal.file.collections.BuildDependenciesOnlyFileCollectionResolveContext.add(BuildDependenciesOnlyFileCollectionResolveContext.java:73)
		at org.gradle.api.internal.file.collections.BuildDependenciesOnlyFileCollectionResolveContext.add(BuildDependenciesOnlyFileCollectionResolveContext.java:83)
		at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.visitContents(DefaultConfigurableFileCollection.java:91)
		at org.gradle.api.internal.file.CompositeFileCollection.visitDependencies(CompositeFileCollection.java:156)
		at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.visitDependencies(DefaultConfigurableFileCollection.java:97)
		at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext$TaskGraphImpl.getNodeValues(CachingTaskDependencyResolveContext.java:88)
		at org.gradle.internal.graph.CachingDirectedGraphWalker$GraphWithEmpyEdges.getNodeValues(CachingDirectedGraphWalker.java:202)
		at org.gradle.internal.graph.CachingDirectedGraphWalker.doSearch(CachingDirectedGraphWalker.java:112)
		at org.gradle.internal.graph.CachingDirectedGraphWalker.findValues(CachingDirectedGraphWalker.java:64)
		at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.doResolve(CachingTaskDependencyResolveContext.java:76)
		at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.resolve(CachingTaskDependencyResolveContext.java:65)
		at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.getDependencies(CachingTaskDependencyResolveContext.java:55)
		at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.addToTaskGraph(DefaultTaskExecutionPlan.java:139)
		at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.addTasks(DefaultTaskGraphExecuter.java:98)
		at org.gradle.execution.TaskNameResolvingBuildConfigurationAction.configure(TaskNameResolvingBuildConfigurationAction.java:47)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.configure(DefaultBuildConfigurationActionExecuter.java:48)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.access$000(DefaultBuildConfigurationActionExecuter.java:25)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter$1.proceed(DefaultBuildConfigurationActionExecuter.java:54)
		at org.gradle.execution.DefaultTasksBuildExecutionAction.configure(DefaultTasksBuildExecutionAction.java:44)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.configure(DefaultBuildConfigurationActionExecuter.java:48)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.access$000(DefaultBuildConfigurationActionExecuter.java:25)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter$1.proceed(DefaultBuildConfigurationActionExecuter.java:54)
		at org.gradle.execution.ExcludedTaskFilteringBuildConfigurationAction.configure(ExcludedTaskFilteringBuildConfigurationAction.java:47)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.configure(DefaultBuildConfigurationActionExecuter.java:48)
		at org.gradle.execution.DefaultBuildConfigurationActionExecuter.select(DefaultBuildConfigurationActionExecuter.java:36)
		at org.gradle.initialization.DefaultGradleLauncher$3.run(DefaultGradleLauncher.java:143)
		at org.gradle.internal.Factories$1.create(Factories.java:22)
		at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
		at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:52)
		at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:140)
		at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
		at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:99)
		at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:93)
		at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
		at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:62)
		at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:93)
		at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:82)
		at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:94)
		at org.gradle.tooling.internal.provider.runner.BuildModelActionRunner.run(BuildModelActionRunner.java:46)
		at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
		at org.gradle.tooling.internal.provider.runner.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:58)
		at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
		at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:43)
		at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:28)
		at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
		at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:48)
		at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:52)
		at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
		at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
		at org.gradle.util.Swapper.swap(Swapper.java:38)
		at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.health.DaemonHealthTracker.execute(DaemonHealthTracker.java:47)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:66)
		at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
		at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.health.HintGCAfterBuild.execute(HintGCAfterBuild.java:41)
		at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
		at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
		at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:246)
		at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
		at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)

**Comment by cmelchior** [comment]
Hmm, that doesn't sound good. It looks like you somehow are using the wrong version of the Transform API. What version of the build tools are you using? And can you post your build.gradle file?

**Comment by roughboy** [comment]
Sure, build tool version is 23.0.2.

Project build.gradle:

	buildscript {
		repositories {
			jcenter()
			maven {
				url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
			}
			maven {
				url "https://jitpack.io"
			}
		}
		dependencies {
			classpath 'com.neenbedankt.gradle.plugins:android-apt:1.8'
			classpath 'com.android.tools.build:gradle:2.0.0-beta6'
			classpath 'io.realm:realm-gradle-plugin:0.88.0-SNAPSHOT'
		}
	}

	allprojects {
		repositories {
			jcenter()
			maven {
				url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
			}
		}
	}


App module build.gradle:

	apply plugin: 'com.android.application'
	apply plugin: 'realm-android'

	android {
		compileSdkVersion 23
		buildToolsVersion "23.0.2"

		defaultConfig {
			applicationId "com.example.lukasz.realmrepro"
			minSdkVersion 16
			targetSdkVersion 23
			versionCode 1
			versionName "1.0"
		}
		buildTypes {
			release {
				minifyEnabled false
				proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
			}
		}
	}

	dependencies {
		compile fileTree(dir: 'libs', include: ['*.jar'])
		testCompile 'junit:junit:4.12'
		compile 'com.android.support:appcompat-v7:23.1.1'
		compile 'com.android.support:design:23.1.1'
	}

**Comment by zaki50** [comment]
I saw this error when I used android gradle plugin 2.0.0-alpha5 or later.
Transform class in 2.0.0-alpha5 or later is not compatible the one in 1.5.0 which Realm in currently using.

To solve this, I think we need to update realm gradle plugin to 2.0.0.

I tried updating android gradle plugin to 2.0.0-beta6. It seems to be working.
https://gist.github.com/zaki50/e907587a1dca9f460beb

**Comment by cmelchior** [comment]
Great, so much for a stable API :cry: Although slightly annoying because wouldn't that force all our users onto 2.*, I am not sure we can depend on 2.* without Gradle forcing that onto everybody using us?

**Comment by zaki50** [comment]
Adding following line to `build.gradle` right after `classpath 'io.realm:realm-gradle-plugin:0.88.0-SNAPSHOT'` seems to solve the problem.

```
configurations.classpath.exclude group: 'com.android.tools.build', module: 'transform-api'
```

Realm Gradle plugin depends on `com.android.tools.build:transform-api:1.5.0`, but it was removed and merged into `com.android.tools.build:gradle` in 2.0.0-alpha5 or later.
So, just removing old dependency solves the problem.

complete `build.gradle` is:
```
buildscript {
    repositories {
        jcenter()
        maven {
            url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
        }
        maven {
            url "https://jitpack.io"
        }
    }
    dependencies {
        classpath 'com.neenbedankt.gradle.plugins:android-apt:1.8'
        classpath 'com.android.tools.build:gradle:2.0.0-beta6'
        classpath 'io.realm:realm-gradle-plugin:0.88.0-SNAPSHOT'

        // please add this
        configurations.classpath.exclude group: 'com.android.tools.build', module: 'transform-api'
    }
}

allprojects {
    repositories {
        jcenter()
        maven {
            url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
        }
    }
}
```

**Comment by zaki50** [comment]
I updated last comment to correct some mistakes.

**Comment by roughboy** [comment]
Just checked it and it just works :) Your user support rocks! Thanks guys!

**Comment by cmelchior** [comment]
While this might solve this particular problem we might run into issues if people use another dependency that depend on the 2.* transform API. I don't think many of those exists (if any), but it still means we most likely cannot use this as an official work-around.

We should probably look deeper into this and figure what the exact difference is and if there is some way we can support both 1.5 and 2.* in our Gradle plugin. However annoying that might be.

**Comment by cmelchior** [comment]
While it would be nice to just bump to 2.*, I am not sure we can force that onto users, especially since 2.* is not stable yet :(

**Comment by zaki50** [comment]
@cmelchior I'm thinking the way to support both 1.5 and 2.0 but currently no idea.

**Comment by zaki50** [comment]
@roughboy happy to hear that. I'll close this issue. Please reopen this if you'll have further problem.
As @cmelchior said, we are seeking the way to support both 1.5 and 2.0.

**Comment by cmelchior** [comment]
Only solution I can come up with is having two implementations of the Transformer and dynamically check which version we need to load

**Comment by zaki50** [comment]
We don't need two implementation(for now), but we need modify dependencies according to the version of Android Gradle plugin.

**Comment by cmelchior** [comment]
@zaki50 In http://oss.jfrog.org/oss-snapshot-local/io/realm/realm-transformer/0.88.0-SNAPSHOT/realm-transformer-0.88.0-20160226.153346-6.pom the transform API is scope: runtime, it should probably be scope: provided. Someone pointed out that while the public API might be compatible the private is not which could cause issues.

**Comment by zaki50** [comment]
@cmelchior right. I checked Android gradle plugin dependencies and found that we didn't need to have runtime dependency from realm-transformer. 

I'll make a PR.

**Comment by zaki50** [comment]
I confirmed that #2355 solves the problem.

**Comment by zaki50** [comment]
I merged #2355 into master.
